File: app/src/test/java/com/orangepenguin/boilerplate/screens/username/UsernamePresenterTest.java
package com.orangepenguin.boilerplate.screens.username;

import com.orangepenguin.boilerplate.model.User;
import com.orangepenguin.boilerplate.singletons.Constants;
import com.orangepenguin.boilerplate.usecase.UserUseCase;
import com.orangepenguin.boilerplate.util.NotificationUtil;
import com.orangepenguin.boilerplate.util.SharedPreferencesUtil;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import rx.Scheduler;
import rx.schedulers.Schedulers;
import rx.schedulers.TestScheduler;
import rx.subjects.PublishSubject;

import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static rx.Observable.just;


/**
 * running with MockitoJUnitRunner enables the use of @Mock annotations
 */
@RunWith(MockitoJUnitRunner.class)
public class UsernamePresenterTest {
    private static final String USERNAME = "testUsername";
    private final TestScheduler scheduler = Schedulers.test(); // observeScheduler to advance time by hand
    @Mock UsernameView mockView;
    @Mock User mockUser;
    @Mock UserUseCase userUseCase;
    @Mock SharedPreferencesUtil sharedPreferencesUtil;
    @Mock NotificationUtil notificationUtil;
    private UsernamePresenter usernamePresenter;

    @Before
    public void setUp() {
        usernamePresenter = new UsernamePresenter(userUseCase, sharedPreferencesUtil, notificationUtil);
        when(userUseCase.fetchUser(USERNAME)).thenReturn(just(mockUser));
    }

    @Test
    public void shouldStartListActivityOnShowUserButtonPress() {
        usernamePresenter.takeView(mockView);
        usernamePresenter.showUserButtonPressed(USERNAME, false);

        verify(mockView).showLoadingIndicator();
        verify(mockView).hideLoadingIndicator();
        verify(mockView).startDetailsActivity(mockUser);
    }

    /**
     * this test illustrates saving a preference
     */
    @Test
    public void shouldSetUsernamePreferenceIfCheckboxChecked() {
        usernamePresenter.takeView(mockView);
        usernamePresenter.showUserButtonPressed(USERNAME, true);
        verify(sharedPreferencesUtil).savePreference(Constants.PREF_USERNAME, USERNAME);
    }

    /**
     * this test illustrates clearing a saved preference
     */
    @Test
    public void shouldClearPreferenceIfCheckboxNotChecked() {
        usernamePresenter.takeView(mockView);
        usernamePresenter.showUserButtonPressed(USERNAME, false);
        verify(sharedPreferencesUtil).clearPreference(Constants.PREF_USERNAME);
    }

    /**
     * this illustrates mocking a saved preference
     */
    @Test
    public void shouldPopulateUsernameAndRememberCheckboxIfUsernamePreferenceSet() {
        when(sharedPreferencesUtil.getPreference(Constants.PREF_USERNAME, null)).thenReturn
                (USERNAME);

        usernamePresenter.takeView(mockView);

        verify(sharedPreferencesUtil).getPreference(Constants.PREF_USERNAME, null);
        verify(mockView).setUsername(USERNAME);
        verify(mockView).checkRememberCheckbox();
    }

    @Test
    public void shouldUpdateLoadingIndicatorOntakeView() {
        // arrange
        setUpDelayedUsersResponse();
        usernamePresenter.takeView(mockView);
        // act
        usernamePresenter.showUserButtonPressed(USERNAME, false);
        // assert
        verify(mockView).showLoadingIndicator();
        // arrange
        UsernameView newMockView = mock(UsernameView.class);
        // Act
        usernamePresenter.takeView(newMockView);
        // Assert
        verify(newMockView).showLoadingIndicator();
        // Act
        scheduler.advanceTimeBy(1500, MILLISECONDS); // User object should now have been returned
        // Assert
        verify(mockView, never()).hideLoadingIndicator();
        verify(newMockView, times(1)).hideLoadingIndicator();
        verify(newMockView).startDetailsActivity(mockUser);
    }

    private void setUpDelayedUsersResponse() {
        Scheduler.Worker worker = scheduler.createWorker(); // worker to schedule events in time
        // Subjects allow both input and output, so they can be swapped in for Observable calls to unit test your code.
        PublishSubject<User> mockUserObservable = PublishSubject.create();
        // schedule an observable event to occur at 1000 ms - return successful response with User object
        worker.schedule(() -> mockUserObservable.onNext(mockUser), 1000, MILLISECONDS);
        worker.schedule(() -> mockUserObservable.onCompleted(), 1000, MILLISECONDS);
        // configure mock client to actually use mock Observable
        when(userUseCase.fetchUser(USERNAME)).thenReturn(mockUserObservable);
    }
}

File: app/src/test/java/com/orangepenguin/boilerplate/screens/username/UsernamePresenterTest.kt
package com.orangepenguin.boilerplate.screens.username

import com.orangepenguin.boilerplate.model.User
import com.orangepenguin.boilerplate.rx.RxTestSchedulers
import com.orangepenguin.boilerplate.singletons.Constants
import com.orangepenguin.boilerplate.usecase.UserUseCase
import com.orangepenguin.boilerplate.util.NotificationUtil
import com.orangepenguin.boilerplate.util.SharedPreferencesUtil
import org.junit.Before
import org.junit.Ignore
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner
import rx.Observable
import rx.Observable.just
import rx.schedulers.Schedulers
import rx.subjects.PublishSubject
import java.util.concurrent.TimeUnit.MILLISECONDS


/**
 * running with MockitoJUnitRunner enables the use of @Mock annotations
 */
@RunWith(MockitoJUnitRunner::class)
class UsernamePresenterTest {
    private val scheduler = Schedulers.test() // observeScheduler to advance time by hand
    @Mock    lateinit var mockView: UsernameView
    @Mock    lateinit var mockUser: User
    @Mock    lateinit var userUseCase: UserUseCase
    @Mock    lateinit var sharedPreferencesUtil: SharedPreferencesUtil
    @Mock    lateinit var notificationUtil: NotificationUtil
    private lateinit var usernamePresenter: UsernamePresenter
    private val schedulers = RxTestSchedulers()

    @Before
    fun setUp() {
        `when`<Observable<User>>(userUseCase.fetchUser(USERNAME)).thenReturn(just<User>(mockUser))
        usernamePresenter = UsernamePresenter(userUseCase, sharedPreferencesUtil, notificationUtil,
                schedulers)
    }

    @Test
    fun shouldStartListActivityOnShowUserButtonPress() {
        usernamePresenter.takeView(mockView)
        usernamePresenter.showUserButtonPressed(USERNAME, false)

        verify<UsernameView>(mockView).showLoadingIndicator()
        verify<UsernameView>(mockView).hideLoadingIndicator()
        verify<UsernameView>(mockView).startDetailsActivity(mockUser)
    }

    /**
     * this test illustrates saving a preference
     */
    @Test
    fun shouldSetUsernamePreferenceIfCheckboxChecked() {
        usernamePresenter.takeView(mockView)
        usernamePresenter.showUserButtonPressed(USERNAME, true)
        verify<SharedPreferencesUtil>(sharedPreferencesUtil).savePreference(Constants.PREF_USERNAME,
                USERNAME)
    }

    /**
     * this test illustrates clearing a saved preference
     */
    @Test
    fun shouldClearPreferenceIfCheckboxNotChecked() {
        usernamePresenter.takeView(mockView)
        usernamePresenter.showUserButtonPressed(USERNAME, false)
        verify<SharedPreferencesUtil>(sharedPreferencesUtil).clearPreference(
                Constants.PREF_USERNAME)
    }

    /**
     * this illustrates mocking a saved preference
     */
    @Test
    @Ignore //TODO: get this mocking nullable business working
    fun shouldPopulateUsernameAndRememberCheckboxIfUsernamePreferenceSet() {
        `when`(sharedPreferencesUtil.getPreference(eq(Constants.PREF_USERNAME),
                ArgumentMatchers.isNull<String>())).thenReturn(USERNAME)

        usernamePresenter.takeView(mockView)

        verify<SharedPreferencesUtil>(sharedPreferencesUtil).getPreference(Constants.PREF_USERNAME,
                ArgumentMatchers.isNull<String>())
        verify<UsernameView>(mockView).setUsername(USERNAME)
        verify<UsernameView>(mockView).checkRememberCheckbox()
    }

    @Test
    fun shouldUpdateLoadingIndicatorOntakeView() {
        // arrange
        setUpDelayedUsersResponse()
        usernamePresenter.takeView(mockView)
        // act
        usernamePresenter.showUserButtonPressed(USERNAME, false)
        // assert
        verify<UsernameView>(mockView).showLoadingIndicator()
        // arrange
        val newMockView = mock(UsernameView::class.java)
        // Act
        usernamePresenter.takeView(newMockView)
        // Assert
        verify(newMockView).showLoadingIndicator()
        // Act
        scheduler.advanceTimeBy(1500, MILLISECONDS) // User object should now have been returned
        // Assert
        verify<UsernameView>(mockView, never()).hideLoadingIndicator()
        verify(newMockView, times(1)).hideLoadingIndicator()
        verify(newMockView).startDetailsActivity(mockUser)
    }

    private fun setUpDelayedUsersResponse() {
        val worker = scheduler.createWorker() // worker to schedule events in time
        // Subjects allow both input and output, so they can be swapped in for Observable calls to unit test your code.
        val mockUserObservable = PublishSubject.create<User>()
        // schedule an observable event to occur at 1000 ms - return successful response with User object
        worker.schedule({ mockUserObservable.onNext(mockUser) }, 1000, MILLISECONDS)
        worker.schedule({ mockUserObservable.onCompleted() }, 1000, MILLISECONDS)
        // configure mock client to actually use mock Observable
        `when`<Observable<User>>(userUseCase.fetchUser(USERNAME)).thenReturn(mockUserObservable)
    }

    companion object {
        private val USERNAME = "testUsername"
    }
}
